Fix various documentation issues

ZedThree opened the pull request:
This is a bid to try and get the number of Sphinx warnings down, to ideally start using error on warnings.

The major issue now is things like:

```
BOUT-dev/manual/sphinx/_breathe_autogen/file/pcr__thomas_8hxx.rst:4: WARNING: Duplicate C++ declaration, also defined at _breathe_autogen/file/pcr__thomas_8hxx:4.
Declaration is '.. cpp:function:: virtual void setCoefA (const Field2D &val)=0'.
```

which I think is complaining about overriding the pure virtual function. I don't know whether this is an issue with breathe or doxygen. Breathe is not super well maintained currently, so I might investigate other options.

I've fixed a massive chunk of these warnings by just not processing the `.cxx` files. I think this is probably reasonable, but it might involve going through and moving docstrings to the headers. That will take a little bit of time to do.

Making a draft now in case anyone has feedback or ideas on these changes.

ZedThree committed:
Docs: Fix some double-backtick issues

ZedThree committed:
Docs: Update code layout

Prefer linking to sections rather than specific files, as these have
changed a fair bit

ZedThree committed:
Docs: Fix underline too short

ZedThree committed:
Docs: Don't include boutdata/boututils in docs

ZedThree committed:
Docs: Fix reference link

ZedThree committed:
Docs: Fix bullet points (require blank line before start)

ZedThree committed:
Docs: Fix wrong start of literal block

ZedThree committed:
Docs: Fix snippet for Scorep

ZedThree committed:
Docs: Fix unknown role

ZedThree committed:
Docs: Fix syntax for footnote

ZedThree committed:
Fix a couple of "most vexing parse" that doxygen doesn't like

ZedThree committed:
Docs: Don't keep separate doxygen config for readthedocs

ZedThree committed:
Remove unused `#define` from petsc/slepc solvers

ZedThree committed:
Remove duplicate function pointer alias

ZedThree committed:
Remove unnecessary type alias

We might want some check that `BoutReal == PetscScalar` but this type
alias would just cover up any problem. At best, this wouldn't compile,
but it's possible it would just lead to link time problems

ZedThree committed:
Docs: Add a couple of macros that doxygen should ignore

ZedThree committed:
Docs: Don't process .cxx files

ZedThree committed:
Docs: Fix some latex equations in docstrings

ZedThree committed:
Tweak `HypreLib` docs

Fixes formatting of example, and ensures file docstring doesn't get
attached to the `bout` namespace and so appear in every single file
that has the namespace!

ZedThree committed:
Docs: Always set breathe config even if not generating files

ZedThree committed:
Docs: Don't show header to include in docs

ZedThree committed:
Docs: Don't run doxygen on python files

ZedThree committed:
Docs: Fix some egregiously outdated solver docs

ZedThree committed:
Docs: Fix some broken links now `.cxx` files aren't included

ZedThree commented:
I've discovered that the vast majority of the warnings emitted from Sphinx are due to two separate bugs in doxygen and breathe.

Things that still need doing:

- [ ] `boutdata/boututils` docs are still built as part of this repo, should probably build them separately
- [ ] go through every single file and make sure the docstring is in the header

ZedThree committed:
Docs: Fix file-level docstrings not rendering correctly

ZedThree committed:
Docs: Fix a bunch of LaTeX/codeblock rendering issues

ZedThree committed:
Apply clang-format changes

dschwoerer commented on manual/sphinx/conf.py:
```diff
@@ -31,7 +31,8 @@
 import sys
 
-sys.path.append("../../tools/pylib")
+sys.path.append("../../tools/pylib/boutpp")
+sys.path.append("../../tools/pylib/boutconfig")
```
Why did you do this?
That breaks the RTD build:
https://bout-dev--2844.org.readthedocs.build/en/2844/_apidoc/boutpp.html#boutpp-api
https://bout-dev.readthedocs.io/en/latest/_apidoc/boutpp.html#boutpp-api

The tools are directly in boutpp and boutconfig ...

ZedThree replied:
This was to stop generating the docs for the other directories under pylib, and to be explicit about what's built here. It looks like it built the docs for everything except these two, which is very odd. Not sure what's going on!

dschwoerer replied:
I think the other ones get pip installed?
I haven't checked though!

Anyway, for zoidberg we currently don't build the docs but reference the bout++ docs, so if you want to drop zoidberg here, we should add zoidberg independently.
Also, it would be better to remove the sections, rather then to have them being empty because python cannot find the modules.

ZedThree commented on include/bout/generic_factory.hxx:
```diff
@@ -1,5 +1,9 @@
 /// Base type for factories
 
+// Copyright 2018-2022 Peter Hill, BOUT++ Team
+// SPDX-License-Identifier: MIT
```
The code is heavily based on Modern C++ Design/Loki by A. Alexandrescu, which is MIT

dschwoerer replied:
Could you please add then a copy of the license?
I think that is even required by the license ...

dschwoerer committed:
Merge branch 'next' of github.com:boutproject/BOUT-dev into fix-docs-warnings

dschwoerer committed:
Add MIT License text

include/bout/generic_factory.hxx is licensed under the MIT License
based on Modern C++ Design/Loki by A. Alexandrescu

bendudson committed:
Merge pull request #2875 from boutproject/fix-docs-warnings-db

Update + add License for #2844 Fix various documentation issues

dschwoerer committed:
Revert "Docs: Don't include boutdata/boututils in docs"

This reverts commit e6bc5c5d0e85b5e895b124b066bbf62aed6c45b1.

The commit tried to make boutuitls and boutdata non-importable for the
documentation tool.  If those modules should be remove from the
documentation, this can be achieved by removing them from
manual/sphinx/_apidoc/. This should not break the documentation fro
boutpp or boutconfig.

bendudson committed:
Merge pull request #2881 from boutproject/fix-docs-warnings-db

Revert "Docs: Don't include boutdata/boututils in docs"

ZedThree committed:
Merge branch 'next' into fix-docs-warnings

* next: (481 commits)
  CI: Show more output of dnf5 to help debugging
  CI: Disable annotations for clang-tidy-review
  CI: Bump clang-tidy-review
  Remove coverage test from github workflow
  CMake: Ensure we find ADIOS2 if we downloaded it
  Use modern `output` API
  Remove unused header include
  Remove unused `x_outer_dirichlet` member from LaplaceXY2 hypre
  Change test hypre type to `gmres`
  Fix laplacexy2-hypre test failing silently
  CMake: Bump downloaded SUNDIALS version
  CMake: Reduce duplication in linking against adios2
  CMake: Ensure we find adios2 when linking against BOUT++
  Apply clang-format changes
  hypre_interface: Add print() method to HypreMatrix
  tests/integrated: Fix hypre tests
  OptionsNetCDF: verifyTimesteps, write override base class
  Testing: Missed 3D + hypre combination
  Disable hypre with 3D metrics
  Prefer scientific notation
  ...

ZedThree committed:
Fix docstring and argument name for `Div_a_Grad_perp`

ZedThree committed:
docs: Fix some links to python docs

ZedThree committed:
docs: Fix a couple of links to no longer included `.cxx` files

ZedThree committed:
Apply black changes

dschwoerer committed:
Merge remote-tracking branch 'origin/next' into fix-docs-warnings

dschwoerer committed:
Fix formatting

dschwoerer committed:
Remove tabs from rst

dschwoerer committed:
Fir formatting and whitespace

dschwoerer committed:
Revert googletest changes

dschwoerer committed:
Merge pull request #3412 from boutproject/fix-docs-warnings-update

Fix docs warnings update

dschwoerer reviewed (dismissed):
Thanks for the cleanup!

dschwoerer committed:
Merge branch 'next' of https://github.com/boutproject/BOUT-dev into fix-docs-warnings

dschwoerer commented on include/bout/fv_ops.hxx:
```diff
@@ -19,59 +19,66 @@
-/*!
- * Div ( a Grad_perp(f) ) -- ∇⊥ ( a ⋅ ∇⊥ f) -- Vorticity
- */
+/// \f[
```
```suggestion
/// Vorticity:
/// \f[
```

dschwoerer committed:
Fixup merge
